src/nks/hcvault/provider.rs

```diff
@@ -0,0 +1,512 @@
+                } else {
+                    println!("Token file does not exist. Generating token...");
+                    // Token field empty and no token in token.json, generate token using API
+                    let runtime = Runtime::new().unwrap();
+                    let nks_address = nks_address.clone().ok_or(SecurityModuleError::NksError)?;
+                    match runtime.block_on(get_token(nks_address.clone())) {
+                        Ok(token) => {
+                            nks_token = token;
+                        }
+                        Err(err) => {
+                            println!("Failed to get tokens from API: {}", err);
+                            return Err(SecurityModuleError::NksError);
+                        }
+                    }
+                }
+            }
+            //store current secrets
+            let runtime = Runtime::new().unwrap();
+            match runtime.block_on(get_secrets(&nks_token.as_str(), &nks_address_str)) {
+                Ok((secrets_json, new_token)) => {
+                    self.secrets_json = Some(secrets_json.parse().unwrap());
+                    nks_token = new_token;
+                }
+                Err(err) => {
+                    println!("Failed to get secrets: {}", err);
+                    return Err(SecurityModuleError::NksError);
+                }
+            }
+            //safe token to config
+            let config = NksConfig::new(
+                nks_token.clone(),
+                nks_config.nks_address.clone(),
+                nks_config.key_algorithm.clone(),
+                nks_config.hash.clone(),
+                nks_config.key_usages.clone(),
+            );
+            self.config = Some(config);
+            //save token in token.json for persistence
+            let token_data = json!({
+                "user_token": nks_token.clone()
+            });
+            fs::write("token.json", token_data.to_string()).expect("Error writing to token.json");
+            println!("Nks initialized successfully.");
+            Ok(())
+        } else {
+            println!("Failed to downcast to NksConfig");
+            Err(SecurityModuleError::NksError)
+        }
+    }
+}
+
+/// Retrieves the user token from the `token.json` file.
+///
+/// This function opens the `token.json` file and reads its contents. It then parses the contents as JSON and retrieves the `user_token` field.
+///
+/// # Returns
+///
+/// An `Option<String>` that, if the file exists and the `user_token` field is found, contains the user token as a `String`.
+/// If the file does not exist, or the `user_token` field is not found, it returns `None`.
+///
+/// # Example
+///
+/// ```
+/// let user_token = get_user_token_from_file();
+/// if let Some(token) = user_token {
+///     println!("User token: {}", token);
+/// } else {
+///     println!("User token not found");
+/// }
+/// ```
+fn get_user_token_from_file() -> Option<String> {
```

Comment:
The mechanism in place to save and load user tokens will likely cause problems when this code is used concurrently.  From what I can see a token is obtained by contacting your web service (API.cs). The token is then saved to a file and read from the file when interaction with Vault required. 

When multiple parties do this, the saved token will be overwritten and two parties will effectively share the same token, which is usually not what you want. Would it not make more sense to return the token once obtained and then pass it in when calling functions needing the token?

Reply:
To clarify, the system is designed for individual users to have their own unique token, and it is not intended for concurrent request processing. Each token is specific to a user and is used to access the cryptographic operations we offer. The requests are expected to be made sequentially. this is now documented in [a8dcfa1](https://github.com/cep-sose2024/rheinsec_rust-crypto/commit/a8dcfa138937c72c2eddecb77109b75d3883dcb7)